### PR TITLE
Python printer

### PR DESCRIPTION
**Pull request by Framba-Luca**

### CODE REVIEW

File: unified_planning/io/python_writer.py [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+
+from fractions import Fraction
+import sys
+import unified_planning as up
+import unified_planning.walkers as walkers
+from unified_planning.model.types import _UserType, _IntType, _RealType
+from typing import IO, Any, Optional, cast
+from io import StringIO
+
+#TODO: remove old commentedcode if it is not needed anymore.
```
Comment (mikand) [inline unified_planning/io/python_writer.py:24]: leftover?

File: unified_planning/io/python_writer.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+    '''This class can be used to write a Problem in PDDL.'''
+
+    def __init__(self, problem: 'up.model.Problem'):
+        self.problem = problem
+
+    def _write_problem_code(self, out: IO[str]):
+
+        converter = ConverterToPythonString(self.problem.env)
+        out.write('import unified_planning as up\n')
+        out.write('from unified_planning.shortcuts import *\n')
```
Comment (mikand) [inline unified_planning/io/python_writer.py:160]: I would avoid this import, you should be able to create everything based on the `env` object and make everything reentrant, no? This makes the generated code much more reusable IMHO

File: unified_planning/io/python_writer.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+            self._write_problem_code(f)
+
+
+def _print_python_type(type: 'up.model.types.Type') -> str:
+    '''This method takes a type and returns how to use it from command line.
+    For the user_types it assumes they have already been created and just refers to them.'''
+    if type.is_user_type():
+        return f'type_{cast(_UserType, type).name()}'
+    elif type.is_bool_type():
+        return 'BoolType()'
```
Comment (mikand) [inline unified_planning/io/python_writer.py:297]: Better:
```suggestion
        return 'env.type_manager.BoolType()'
```

File: unified_planning/io/python_writer.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+        return f'RealType({str(cast(_RealType, type).lower_bound())}, {str(cast(_RealType, type).upper_bound())})'
+    else:
+        raise NotImplementedError
+
+def _convert_timing(timing: up.model.Timing) -> str:
+    bound: str = f'{str(timing.bound())}'
+    if isinstance(timing.bound(), Fraction):
+        bound = f'Fraction({str(timing.bound().numerator)}, {str(timing.bound().denomiantor)})'
+    if timing.is_from_start():
+        return f'StartTiming({bound})'
```
Comment (mikand) [inline unified_planning/io/python_writer.py:310]: I think you need to import these explicitly or use the full name `up.model.timing.StartTiming`

**Review by mikand — CHANGES_REQUESTED** [review]
@Framba-Luca very nice! I left just some minor commits

File: unified_planning/io/python_writer.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+        converter = ConverterToPythonString(self.problem.env)
+        out.write('import unified_planning as up\n')
+        out.write('from unified_planning.shortcuts import *\n')
+        out.write('env = up.environment.get_env()\n')
+        out.write('emgr = env.expression_manager\n')
+        out.write('tm = env.type_manager\n')
+
+        for type in self.problem.user_types(): # define user_types
+            if cast(_UserType, type).father() is None:
+                out.write(f'type_{cast(_UserType, type).name()} = tm.UserType("{cast(_UserType, type).name()}")\n')
```
[the remark this answers is not in the document]
Comment (Framba-Luca) [reply unified_planning/io/python_writer.py:167]: @mikand  I would love to, and it was my first try and also my hope. But it doesn't work :cry:

File: unified_planning/io/python_writer.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+                out.write('up.model.metrics.MinimizeActionCosts()')
+            elif isinstance(qm, up.model.metrics.MinimizeExpressionOnFinalState):
+                out.write(f'up.model.metrics.MinimizeExpressionOnFinalState({converter.convert(qm.expression)})')
+            elif isinstance(qm, up.model.metrics.MaximizeExpressionOnFinalState):
+                out.write(f'up.model.metrics.MaximizeExpressionOnFinalState({converter.convert(qm.expression)})')
+            else:
+                raise NotImplementedError
+            out.write(')\n')
+
+    def print_domain(self): # TODO: find a better name
```
Comment (alvalentini) [inline unified_planning/io/python_writer.py:257]: This TODO should be removed or fixed

File: unified_planning/io/python_writer.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+        '''Prints to std output the PDDL domain.'''
+        self._write_problem_code(sys.stdout)
+
+    def write_problem_code(self) -> str:
+        '''Returns the PDDL domain.'''
+        out = StringIO()
+        self._write_problem_code(out)
+        return out.getvalue()
+
+    def write_domain(self, filename: str): # TODO: find a better name
```
Comment (alvalentini) [inline unified_planning/io/python_writer.py:267]: Same here

File: unified_planning/io/python_writer.py [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+            elif isinstance(qm, up.model.metrics.MinimizeExpressionOnFinalState):
+                out.write(f'up.model.metrics.MinimizeExpressionOnFinalState({converter.convert(qm.expression)})')
+            elif isinstance(qm, up.model.metrics.MaximizeExpressionOnFinalState):
+                out.write(f'up.model.metrics.MaximizeExpressionOnFinalState({converter.convert(qm.expression)})')
+            else:
+                raise NotImplementedError
+            out.write(')\n')
+
+    def print_domain(self): # TODO: find a better name
+        '''Prints to std output the PDDL domain.'''
```
Comment (alvalentini) [inline unified_planning/io/python_writer.py:258]: Is it PDDL?

File: unified_planning/io/python_writer.py [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+        self._write_problem_code(sys.stdout)
+
+    def write_problem_code(self) -> str:
+        '''Returns the PDDL domain.'''
```
Comment (alvalentini) [inline unified_planning/io/python_writer.py:262]: Is it PDDL?

File: unified_planning/model/problem.py [hunk 10]
```diff
@@ -471,6 +473,19 @@ def initial_values(self) -> Dict['up.model.fnode.FNode', 'up.model.fnode.FNode']
...
+    
+    def initial_defaults(self) -> Dict['up.model.types.Type', 'up.model.fnode.FNode']:
+        '''Returns the problem's initial defaults.'''
+        return self._initial_defaults
+    
+    def fluents_defaults(self) -> Dict['up.model.fluent.Fluent', 'up.model.fnode.FNode']:
+        '''Returns the problem's fluents defaults.'''
+        return self._fluents_defaults
+    
+    def _initial_values_structure(self) -> Dict['up.model.fnode.FNode', 'up.model.fnode.FNode']:
```
Comment (alvalentini) [inline unified_planning/model/problem.py:485]: This method is used outside the `Problem` class. We should remove the initial `_`.

I think also that `explicit_initial_values` could be a better name. What do you think?

**Review by alvalentini — CHANGES_REQUESTED** [review]
@Framba-Luca  I added some minor comments.

File: unified_planning/io/python_writer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mikand) [reply unified_planning/io/python_writer.py:149]: @Framba-Luca I tried the following and it seems to work on mypy:
```diff
diff --git a/unified_planning/io/python_writer.py b/unified_planning/io/python_writer.py
index 0621ffb..94571e9 100644
--- a/unified_planning/io/python_writer.py
+++ b/unified_planning/io/python_writer.py
@@ -145,10 +145,11 @@ class PythonWriter:
         out.write('tm = env.type_manager\n')
 
         for type in self.problem.user_types(): # define user_types
-            if cast(_UserType, type).father() is None:
-                out.write(f'type_{cast(_UserType, type).name()} = tm.UserType("{cast(_UserType, type).name()}")\n')
+            utype = cast(_UserType, type)
+            if utype.father() is None:
+                out.write(f'type_{utype.name()} = tm.UserType("{utype.name()}")\n')
             else:
-                out.write(f'type_{cast(_UserType, type).name()} = tm.UserType("{cast(_UserType, type).name()}", type_{cast(_UserType, cast(_UserType, type).father()).name()})\n') 
+                out.write(f'type_{utype.name()} = tm.UserType("{utype.name()}", type_{cast(_UserType, utype.father()).name()})\n') 
 
         for f in self.problem.fluents(): # define fluents
             params = ', '.join(_print_python_type(p) for p in f.signature())
```

File: unified_planning/io/python_writer.py [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+                if e.is_increase():
+                    out.write(f'problem.add_increase_effect(timing={_convert_timing(t)}, fluent={converter.convert(e.fluent())}, value={converter.convert(e.value())}, condition={converter.convert(e.condition())})\n')
+                elif e.is_decrease():
+                    out.write(f'problem.add_decrease_effect(timing={_convert_timing(t)}, fluent={converter.convert(e.fluent())}, value={converter.convert(e.value())}, condition={converter.convert(e.condition())})\n')
+                else:
+                    out.write(f'problem.add_timed_effect(timing={_convert_timing(t)}, fluent={converter.convert(e.fluent())}, value={converter.convert(e.value())}, condition={converter.convert(e.condition())})\n')
+        
+        for i, gl in self.problem.timed_goals().items(): # add timed goals
+            for g in gl:
+                out.write(f'problem.add_timed_goal(timing={_convert_interval(i)}, goal={converter.convert(g)})\n')
```
Comment (alvalentini) [inline unified_planning/io/python_writer.py:234]:
```suggestion
                out.write(f'problem.add_timed_goal(interval={_convert_interval(i)}, goal={converter.convert(g)})\n')
```
or
```suggestion
                out.write(f'problem.add_timed_goal({_convert_interval(i)}, {converter.convert(g)})\n')
```

**Review by mikand — APPROVED** [review]

**Pull request merged**